### PR TITLE
Sandip devlopement

### PR DESCRIPTION
**Pull request by nirmalsandip**
Developement for Loan TP and Payment TP

### CODE REVIEW

File: sawtooth-asset-app/src/main/java/com/mycompany/blockchain/sawtooth/app/config/Application.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
  *
  */
 @SpringBootApplication
-@ComponentScan("com.mycompany.blockchain.sawtooth.app")
+@ComponentScan("com.mycompany.blockchain.sawtooth")
```
Comment (nissshh) [inline sawtooth-asset-app/src/main/java/com/mycompany/blockchain/sawtooth/app/config/Application.java]: Put component scan at app level, other packages are base package classes and we will let applicaitons to define the beans. Please add .app.

File: sawtooth-asset-app/src/main/resources/application.properties [hunk 2]
```diff
@@ -1,4 +1,7 @@
 asset.tx.family.name=asset
+loan.tx.family.name=loan
+wallet.tx.family.name=wallet
+payment.tx.family.name=payment
 asset.tx.family.version=1.0
```
Comment (nissshh) [inline sawtooth-asset-app/src/main/resources/application.properties]: add separate versionsfor transaction family.

File: sawtooth-java-wrapper-client/src/test/java/com/mycompany/blockchain/sawtooth/client/loan/LoanPayloadClientTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+public class LoanPayloadClientTest extends BaseClientTest {
+
+	Logger logger = Logger.getLogger(LoanPayloadClientTest.class.getName());
+
+	String assetId = "A001", borrowerId = "B001", lenderId = "L001";
+
+	@Override
+	protected ClientService getClientService() {
+		return new LoanPayloadClientService("loan", "1.0", null, getZMQAddress());
```
Comment (nissshh) [inline sawtooth-java-wrapper-client/src/test/java/com/mycompany/blockchain/sawtooth/client/loan/LoanPayloadClientTest.java]: Create a test user keyu in stead of null to let the user key testing also done.

File: sawtooth-java-wrapper-model/src/main/resources/protos/loan.proto [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+     	string asset_id = 1;
+     	
+     	//Payment Details
+     	Payment payment = 2;
+     }
+        
+    enum PayloadType {        
+        CREATE_LOAN = 0;
+        APPROVE_LOAN = 1;  
+        MONTHLY_PAYMENT = 2;
```
Comment (nissshh) [inline sawtooth-java-wrapper-model/src/main/resources/protos/loan.proto]: use generic like REPAYMENT

File: sawtooth-asset-app/src/main/java/com/mycompany/blockchain/sawtooth/app/service/LoanService.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		TransferPayment transferPayment = TransferPayment.newBuilder()
+				.setAmount(payment.getAmount()).setDestCustomerId(payment.getTo())
+				.setSourceCustomerId(payment.getFrom()).build();
+		walletService.transferWallet(transferPayment);
+		
+		//Loan TP
+		LoanRequestPayload loanRequestPayload = LoanRequestPayload.newBuilder().setPayloadType(PayloadType.MONTHLY_PAYMENT)
+				.setMonthlyPayment(loanPaymentPayload).build();
+		log.info("Sending Loan Payment Payload as " + loanPaymentPayload);
+		response = service.submitStateChange(loanRequestPayload);
```
Comment (nissshh) [inline sawtooth-asset-app/src/main/java/com/mycompany/blockchain/sawtooth/app/service/LoanService.java]: create a new branch to implement the complete transactions for payment, wallet and loan updation.

File: sawtooth-java-wrapper-service/src/main/java/com/mycompany/blockchain/sawtooth/core/service/payment/PaymentValidator.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see com.mycompany.blockchain.sawtooth.core.service.IValidator#validate(java.lang. Object)
+	 */
+	@Override
+	public boolean validate(Payment data) throws InvalidTransactionException {
+		if (data.getId() == null || data.getId().isEmpty())
+			return false;
```
Comment (nissshh) [inline sawtooth-java-wrapper-service/src/main/java/com/mycompany/blockchain/sawtooth/core/service/payment/PaymentValidator.java]: Throw exception here and not a false type. the message should describe the error.

File: sawtooth-java-wrapper-service/src/main/java/com/mycompany/blockchain/sawtooth/core/service/loan/LoanHandler.java [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+		return Loan.newBuilder().setAssetId(existingLoan.getAssetId()).setBorrowerId(existingLoan.getBorrowerId())
+				.setLenderId(existingLoan.getLenderId()).setRequestedAmt(existingLoan.getRequestedAmt())
+				.setApprovedAmt(existingLoan.getApprovedAmt()).setId(existingLoan.getId()).setBalance(balance)
+				.setRoi(existingLoan.getRoi()).setStatus(existingLoan.getStatus()).addAllPayments(paymentsList)
+				.build();
+	}
+
+	@Override
+	public IAddressBuilder<LoanRequestPayload> getAddressBuilder() {
+		return null;
```
Comment (nissshh) [inline sawtooth-java-wrapper-service/src/main/java/com/mycompany/blockchain/sawtooth/core/service/loan/LoanHandler.java]: dont send null.return the address bulders and dao.

File: sawtooth-asset-app/src/main/java/com/mycompany/blockchain/sawtooth/app/config/Application.java (hunk 1 above)
Comment (nirmalsandip) [reply sawtooth-asset-app/src/main/java/com/mycompany/blockchain/sawtooth/app/config/Application.java]: done

File: sawtooth-asset-app/src/main/resources/application.properties (hunk 2 above)
Comment (nirmalsandip) [reply sawtooth-asset-app/src/main/resources/application.properties]: done

File: sawtooth-java-wrapper-model/src/main/resources/protos/loan.proto (hunk 4 above)
Comment (nirmalsandip) [reply sawtooth-java-wrapper-model/src/main/resources/protos/loan.proto]: done

File: sawtooth-java-wrapper-service/src/main/java/com/mycompany/blockchain/sawtooth/core/service/payment/PaymentValidator.java (hunk 6 above)
Comment (nirmalsandip) [reply sawtooth-java-wrapper-service/src/main/java/com/mycompany/blockchain/sawtooth/core/service/payment/PaymentValidator.java]: Done

File: sawtooth-java-wrapper-client/src/test/java/com/mycompany/blockchain/sawtooth/client/loan/LoanPayloadClientTest.java [hunk 8]
```diff
@@ -109,4 +112,26 @@ public void testReadData() throws UnsupportedEncodingException, InvalidProtocolB
...
+				.setBorrowerId(borrowerId).setLenderId(lenderId).build();
+		String address = new LoanAddressBuilder("loan", "1.0").buildAddress(loan);
+				
+		List<Entry> responses = zmqTemplate.getClientListStateRequest(address.substring(0,6));
+		List<Loan> allLoans = new ArrayList<>();
+		responses.forEach(entry -> {
+			try {
+				allLoans.add(Loan.parseFrom(entry.getData()));
+			} catch (InvalidProtocolBufferException e) {
+				System.out.println(e);
```
Comment (nissshh) [inline sawtooth-java-wrapper-client/src/test/java/com/mycompany/blockchain/sawtooth/client/loan/LoanPayloadClientTest.java]: use a logger hear always , practice to add a logger always.

**Pull request merged**